agent/consul/state/catalog.go

```diff
@@ -1050,6 +1073,13 @@ func (s *Store) deleteServiceTxn(tx *memdb.Txn, idx uint64, nodeName, serviceID
 	}
 
 	svc := service.(*structs.ServiceNode)
+	// In some cases, it would be possible to avoid updating this index if
+	// the tags of this service instance are all present in other instances
+	// But the optimization might be costly since we would have to run thru the
+	// whole list of services.
```

Comment:
This comment a tiny bit misleading - even if there were no unique tags on the service that is deregistering, the /catalog/services endpoint output might still need to change if this is the last instance of the named service. We already check that below from your previous PR but I agree we should just always update here for now as it's simpler to reason about and we'd need to check tags too like you said.

Could we just change the comment to include that case that last-service instance would still need to bump this?

Reply:
Hello Paul,

Regarding eTags and friends, I could not agree more, I even proposed this a few weeks ago: https://github.com/hashicorp/consul/issues/3871

Regarding the eTag, there are a few issues however, since for instance the order is not deterministic for now (thus basic shasum or similar cannot be done on string value for instance - but of course definitely solveable)

Well, the main optimisation was definitely on the index per service. This is a bit the cherry on the icing, but I still think it worth the optimisation especially for consul-template like tools (and in our case, prometheus).

Discuss this internally, but the first optimisation (per service name really "destroyed" all of our scalability issues with Consul - agreed that we might be using Consul differently from several users, but still), and this one might finish the Job.

This PR also do not handle all cases (it does not optimize the case where node meta is used for filtering), so allowing to optimize only common use-cases is a first nice step, but it is unlikely that this optimisation might be used for something else than /catalog/services for now, thus the comment. I update my comment in the source code right away. All of this is purely internal, so I doubt this kind of optimisation cannot be removed or changed to something else in the future anyway.